Add WireGuard host2host and LB encryption

brb opened the pull request:
This PR adds support for node-to-node encryption to WireGuard. To achieve this, we've completely changed the WireGuard integration in the datapath. Previously, WireGuard support was implemented by marking packets to be encrypted in `"from-container"` and redirecting it to the WireGuard tunnel via a hostns IP rule. This worked fine for traffic originating in pods - but for node-to-node traffic, we need to redirect the packets on the outgoing network interface. Thus, the new implementation attaches `bpf_host` to the outgoing device and redirects packets to the WireGuard tunnel from there. See commit descriptions for more details.

On the agent side, there are also changes to the implementation. Previously, the datapath assumed that any IPCache entry with an associated tunnel endpoint would need encryption. To determine if we need to encryption traffic to a remote endpoint, we now rely on the `encrypt_key` field instead. This allows us to more precisely track if traffic to a particular destination needs to be encrypted, and allows certain nodes to opt out of encryption (see below). The agent code has been updated to populate the CiliumEndpoint and CiliumNode CRDs with a static non-zero `EncryptKey` value if encryption for those resources is enabled.

Additional points worth noting:
 - This new implementation now supports L7 proxy with WireGuard (due to it not requiring skb marks pre-encryption)
 - The agent opts out of node-to-node encryption if it is running on a kube-apiserver control plane node. This is done to
ensure worker nodes are always able to communicate with the kube-apiserver node, in order to be able to manage their own encryption key. See docs for more details.
 - CI is handled in the new CI-Datapath suite via Cilium-CLI. The run with the new testsuite (c.f. https://github.com/cilium/cilium-cli/pull/1308) was successful: 
   :heavy_check_mark:  https://github.com/cilium/cilium/actions/runs/3884806887/jobs/6627856819

:information_source:  **Please see commit messages for many more details**

Joint work between @gandro and @brb

Follow-ups, to be done in separate PRs:
- [ ] (follow-up) Ability to store private key persistently (as an alternative to opting out of node-to-node encryption for control-plane nodes)

gandro committed:
node: Set EncryptKey if WireGuard is enabled

This commit sets the EncryptKey index field to a static non-zero value
in CiliumEndpoints and CiliumNodes if WireGuard is enabled for managing
cilium-agent. This allows us to have fine-grained tracking of encrypted
endpoints. A follow-up commit will update the datapath accordingly, to
make use of this field and only encrypt connections to endpoints which
have a non-zero encryption key index.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

gandro committed:
helm: Enable node encryption for WireGuard

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Drop L7 check for WireGuard

There is no longer skb mark conflict with L7 proxy, so we can drop the
check. This means that the L7 proxy can work together with the WG
integration.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Detect bpf_host netdevs when WG is enabled

We need to detect a direct routing dev (= one which is used to connect
K8s Nodes) in order to attach bpf_host when WG is enabled, as bpf_host
is responsible for redirecting packets to the WG netdev for encryption.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
datapath: Change WG integration to support host2host case

This commit completely changes the WireGuard integration in the
datapath to enable the host2host encryption (also, pod2host and
host2pod).

Previously, we supported only the pod2pod case. This was implemented by
marking a to be encrypted packet, and then letting the IP rule installed
in the host netns to forward the packet to the WireGuard tunnel device
"cilium_wg0" for the encryption, as shown below:

┌─────────────────┐
│   Pod A netns   │
│   ┌────────┐    │
│   │  eth0  │    │
└───┴────┬───┴────┘
    ┌────┴──────────┐
    │ bpf_lxc@veth0 │               (host netns)
    └────┬──────────┘
         │1."from-container" in bpf_lxc sets MARK_MAGIC_ENCRYPT
         │2. ip rule matches the mark and routes packet to WG netdev
         │       ┌───────────┐
         └──────►│cilium_wg0 │
                 └────┬──────┘
                      │
                  ┌───▼───┐
                  │ eth0  │
                  └───────┘

This was working fine for the pod2pod case (albeit one danger that a
sysadmin could nuke the rule making the packet to bypass the WG dev).

However, with this approach it was not possible to enable the host2host
case, as a packet originating from the host netns was never handled by
bpf_lxc. Thus, we needed to change the datapath.

To encrypt a host2host packet we need to attach bpf_host to the outgoing
device connecting cluster nodes which in the picture is "eth0". Then the
program "to-netdev" from bpf_host can forward the packet to the WG dev.
Once encrypted, the packet will be again hitting the same bpf_host
program. To avoid the packet looping forever, we can configure the WG
netdev to set the skb mark after the encryption. Then, in the program
we can skip the redirection to the WG netdev if the mark is set.

The flow below shows the new integration.

┌─────────────────┐
│   Pod A netns   │
│   ┌────────┐    │
│   │  eth0  │    │
└───┴────┬───┴────┘
    ┌────┴──────────┐
    │ bpf_lxc@veth0 │    (host netns)
    └────┬──────────┘
         │
    ┌────▼───────────┐  1. "to-netdev" does redirect    ┌───────────┐
    │ bpf_host@eth0  │─────────────────────────────────►│cilium_wg0 │
    └─┬─────────▲────┘                                  └──────┬────┘
      │         │                                              │
      │         │  2. encrypt and set MARK_MAGIC_ENCRYPTED     │
      │         └──────────────────────────────────────────────┘
      │
      │ 3. output the encrypted packet
      │
      ▼

The same flow is used for the host2host, host2pod and pod2host cases.

Another advantage of this change is that the WG can be used with the L7
proxy (the mutual exclusion check is going to be removed in a subsequent
commit).

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Adopt agent to new WireGuard integration

This commit changes the agent code to support the new WireGuard
integration described in the previous commit.

The most important changes:

1. Configure the WG netdev to add the skb mark.
2. Add NodeIP to allowed-ips when --encrypt-node=true.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Enable node encryption for WG tests with vxlan

Currently, the bpf_overlay prog doesn't redirect a packet to the WG
netdev for encryption (will be addressed in a follow-up PR). So, in
order for the tests to pass, we need to enable the host2host encryption.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
datapath: Bring back WG pod2pod behavior when tunnel != disabled

Before changing the WG integration's behavior, when running in the
tunneling mode, a pod2pod@remote-node traffic escaped the bpf_overlay's
tunneling, and was encapsulated once by the WG tunnel netdev.

To be compatible with this < v1.13 behavior, this commit adds the
redirect to the WG tunnel to the __encap_and_redirect_with_nodeid()
function which is eventually called in the pod2pod packet path.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Encrypt NodePort BPF traffic with WireGuard

This commit attaches the bpf_host's "from-netdev" section to the
Cilium's WireGuard tunnel netdev ("cilium_wg0").

This is needed to enable the encryption of the KPR traffic. In
particular, we encrypt the N/S KPR requests which will be forwarded to
a remote node running a selected service endpoint.

IMPORTANT: this encrypts KPR traffic only when running in the
non-tunneling mode.

For the request path no changes are required. The existing datapath
configuration already handles it, as shown in the following:

1. The "from-netdev" attached to eth0 is invoked for the NodePort
   request.
2. A remote service endpoint is selected, the DNAT and SNAT translations
   are performed.
3. The translated request is redirected to eth0.
4. The "to-netdev" section on eth0 is invoked. It detects that the
   packet needs to encrypted, so it redirects to the cilium_wg0.

For the reply path a minimal changes were required. After the WG netdev
has decrypted the reply packet, the packet is returned to the networking
stack. Because the networking stack is not aware of the connection, the
reply packet is dropped. To avoid that, we attach the "from-netdev"
section to the WG netdev, so that the following can be performed:

1. Reverse SNAT and DNAT translations are applied to the reply.
2. The reply packet is redirected to the outgoing interface.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

gandro committed:
daemon: Make control-plane nodes opt-out of node-to-node encryption

This commit introduces a new command-line option to specify a label
selector to make nodes opt-out of node-to-node encryption. The default
label selector set will match kubeadm control-plane nodes (i.e. the
nodes hosting kube-apiserver). This ensures that all Cilium-managed
nodes will be able to reach the kube-apiserver running on that node
regardless of encryption status. This is important, because we want to
ensure that nodes can change their public keys when they re-join the
cluster.

Nodes who opted out of node-to-node encryption will still perform
encryption for pod-to-pod traffic.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

tommyp1ckles commented on bpf/lib/common.h:
```diff
@@ -575,6 +575,13 @@ enum metric_dir {
  */
 #define MARK_MAGIC_SNAT_DONE		0x1500
 
+/* The mark is used to indicate that the WireGuard tunnel device has done
```
nit: "device has done" -> "device is done"

gandro replied:
Fixed!

tommyp1ckles commented on pkg/wireguard/agent/agent.go:
```diff
@@ -654,3 +640,37 @@ func (p *peerConfig) insertAllowedIP(ip net.IPNet) (updated bool) {
 	return true
 }
+
+// Removes < v1.12 IP rules and routes.
```
This [comment](https://github.com/cilium/cilium/pull/19401/commits/bd527849e6cdbf5b605c89c82ae2c4dea4e1c08a#diff-33d5a29cc08b3cb0d4e53cd4d5bd677a32f4c60bf1b4eb4e2856efb5bd51280aR250) indicates that this removes <1.13 rules.

gandro replied:
Thanks, fixed!

tommyp1ckles commented on pkg/wireguard/agent/agent.go:
```diff
@@ -654,3 +640,37 @@ func (p *peerConfig) insertAllowedIP(ip net.IPNet) (updated bool) {
+		Table:  linux_defaults.RouteTableWireguard,
+	}
+	if option.Config.EnableIPv4 {
+		route.DeleteRule(rule)
```
`DeleteRule`,`Delete`,`DeleteRuleIPv6` all return an error, should we ignoring those?

tommyp1ckles commented on bpf/lib/wireguard.h:
```diff
@@ -0,0 +1,78 @@
+	 * required.
+	 *
+	 * After the packet has been encrypted, the WG tunnel device
+	 * will set the MARK_MAGIC_WG_ENCRYPTED skb mark. So, to avoid
```
I'm confused how this gets set, is it purely in the wg implementation or do we have a bpf program on wg interface?

pchaigno replied:
Same. It's only defined and read in this commit, never written.

pchaigno replied:
That actually happens in a later commit. Grep for `FirewallMark`. But yeah, this character being used in the story before it's introduced leaves the reader somewhat confused :sweat_smile:

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -59,6 +59,10 @@
 
+#ifdef ENABLE_WIREGUARD
+#include "lib/wireguard.h"
+#endif
```
Could we move the `ENABLE_WIREGUARD` guard to the header file and thus avoid having it in every file that includes that header file? Like we do for other header files.

pchaigno commented on bpf/Makefile:
```diff
@@ -180,7 +180,7 @@ HOST_OPTIONS = $(LXC_OPTIONS) \
 
 ifndef MAX_HOST_OPTIONS
-MAX_HOST_OPTIONS = $(MAX_BASE_OPTIONS) -DENCAP_IFINDEX=1 -DTUNNEL_MODE=1
+MAX_HOST_OPTIONS = $(MAX_BASE_OPTIONS) -DENCAP_IFINDEX=1 -DTUNNEL_MODE=1 -DENABLE_WIREGUARD=1
```
Isn't another consequence of commit `datapath: Change WG integration to support host2host case` that L7 proxy redirects are now supported? Probably worth adding a note in the commit description if that's the case, no?

brb replied:
Yep. It's mentioned in the ` daemon: Drop L7 check for WireGuard` commit msg.

pchaigno replied:
Yep but the other commit message doesn't say when it was fixed either.

brb replied:
Added to the `datapath: Change WG integration to support host2host case` commit.

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -1186,6 +1190,25 @@ int cil_to_netdev(struct __ctx_buff *ctx __maybe_unused)
+	 * cluster.
+	 * Once the assumption is no longer true, we will need to recirculate
+	 * the packet back to the "to-netdev" section for the SNAT instead of
+	 * returning TC_ACT_REDIRECT.
```
Same is true today for SRv6, but fairly likely to change in the future.
Is the same true for the health checks block below as well?

brb replied:
> Is the same true for the health checks block below as well?

Yep. These are checks to probe a service backends, so a LB node -> backend node packet should be never masqueraded.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -539,11 +530,7 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
-	if (dst_remote_ep)
-		set_encrypt_mark(ctx);
-	else
-#elif !defined(TUNNEL_MODE)
```
Don't forget to remove the `ENABLE_WIREGUARD` in the comment below.

brb replied:
Good eye.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -1115,11 +1084,7 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
-	if (dst_remote_ep)
-		set_encrypt_mark(ctx);
-	else /* Wireguard and identity mark are mutually exclusive */
-#elif !defined(TUNNEL_MODE)
```
Don't forget to remove the `ENABLE_WIREGUARD` in the comment below.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -241,11 +240,6 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
 			*dst_id = info->sec_label;
 			tunnel_endpoint = info->tunnel_endpoint;
 			encrypt_key = get_min_encrypt_key(info->key);
-#ifdef ENABLE_WIREGUARD
```
AFAICS, all uses of `ENABLE_WIREGUARD` have been removed from bpf_lxc, so we should be able to remove `ENABLE_WIREGUARD` from the `bpf/complexity-tests/*/bpf_lxc.txt` files as well as from `MAX_LXC_OPTIONS`, no?

brb replied:
Aye, good point :pizza:

pchaigno commented on bpf/lib/common.h:
```diff
@@ -575,6 +575,13 @@ enum metric_dir {
+ * fine, as it's not used by K8s. See pkg/datapath/linux/linux_defaults/mark.go
+ * for more details.
+ */
+#define MARK_MAGIC_WG_ENCRYPTED		0x1E00
```
Should we add this mark to the `pkg/datapath/linux/linux_defaults/mark.go` file even if it's not used by the agent?

brb replied:
I think it's already there:
```
	// MagicMarkWireGuardEncrypted is set by the WireGuard tunnel device
	// in order to indicate that a packet has been encrypted, and that there
	// is no need to forward it again to the WG tunnel netdev.
	//
	// The mark invades the K8s mark space described above. This is because
	// some packets might carry a security identity which is indicated with
	// MagicMarkIdentity which takes all 4 bits. The LSB bit which we take
	// from the K8s space is not used, so this is fine).
	MagicMarkWireGuardEncrypted int = 0x1E00
```

pchaigno replied:
Yep, added in a subsequent commit...

pchaigno commented on bpf/lib/common.h:
```diff
@@ -575,6 +575,13 @@ enum metric_dir {
 
+/* The mark is used to indicate that the WireGuard tunnel device has done
+ * encrypting a packet. The MSB invades the Kubernetes mark "space" which is
+ * fine, as it's not used by K8s. See pkg/datapath/linux/linux_defaults/mark.go
```
I'm not sure I understand this sentence:

> The MSB invades the Kubernetes mark "space" which is fine, as it's not used by K8s.

So is it or is it not used by Kubernetes?

brb replied:
> I'm not sure I understand this sentence:

Neither me :sweat_smile: 

What we wanted to say is that we are taking one bit from the K8s mark "space", and that bit is not used by K8s. I.e., the bit is `0x1000`, and the K8s marks are `0x4000` and `0x8000`. So both are not interfering with that bit.

brb replied:
Do you want me to update the comment to make it more clear?

pchaigno replied:
Yes please. I would just mention the specific values as you just did here.

pchaigno commented on bpf/lib/wireguard.h:
```diff
@@ -0,0 +1,78 @@
+	     /* host2pod */ && (src && src->sec_label != HOST_ID)
+#endif /* ENABLE_NODE_ENCRYPTION */
+	)
+		return ctx_redirect(ctx, WG_IFINDEX, 0);
```
Where is `WG_IFINDEX` set?

EDIT: In a latter commit...

pchaigno replied:
There are three cases in that single condition/single comment. Why not make that apparent?
```suggestion
	/* After the packet has been encrypted, the WG tunnel device
	 * will set the MARK_MAGIC_WG_ENCRYPTED skb mark. So, to avoid
	 * looping forever (e.g., bpf_host@eth0 => cilium_wg0 =>
	 * bpf_host@eth0 => ...; this happens when eth0 is used to send
	 * encrypted WireGuard UDP packets), we check whether the mark
	 * is set before the redirect.
	 */
	 if ((ctx->mark & MARK_MAGIC_WG_ENCRYPTED) ==
	     MARK_MAGIC_WG_ENCRYPTED)
		goto out;

	/* Unless node encryption is enabled, we don't want to encrypt
	 * traffic from the hostns.
	 */
#ifndef ENABLE_NODE_ENCRYPTION
	if (!src || src->sec_label == HOST_ID)
		goto out;
#endif /* ENABLE_NODE_ENCRYPTION */

	/* Redirect to the WireGuard tunnel device if the encryption is
	 * required.
	 */
	if (dst && dst->key)
		return ctx_redirect(ctx, WG_IFINDEX, 0);
```

pchaigno commented on bpf/lib/wireguard.h:
```diff
@@ -0,0 +1,78 @@
+	if (((ctx->mark & MARK_MAGIC_WG_ENCRYPTED) !=
+	     MARK_MAGIC_WG_ENCRYPTED) && (dst && dst->key)
+#ifndef ENABLE_NODE_ENCRYPTION
+	     /* host2pod */ && (src && src->sec_label != HOST_ID)
```
Hm, this is a bit dangerous and probably worth a comment. In case KPR=disabled, kube-proxy may have performed masquerading on packets we process here, so the `HOST_ID` source identity may be incorrect. I believe we're fine in our case because we are only interested in pod-to-pod traffic which shouldn't be masqueraded, but it's probably worth a comment at least.

brb replied:
Added the note.

pchaigno commented on bpf/lib/wireguard.h:
```diff
@@ -0,0 +1,78 @@
+#endif
+#ifdef ENABLE_IPV4
+	struct iphdr *ip4;
+#endif
```
Nit: Could we just use `__maybe_unused` for such variables, to avoid the many #ifdefs.

pchaigno commented on pkg/option/config.go:
```diff
@@ -2599,7 +2599,7 @@ func (c *DaemonConfig) DirectRoutingDeviceRequired() bool {
 	// BPF NodePort and BPF Host Routing are using the direct routing device now.
 	// When tunneling is enabled, node-to-node redirection will be done by tunneling.
 	BPFHostRoutingEnabled := !c.EnableHostLegacyRouting
-	return (c.EnableNodePort || BPFHostRoutingEnabled) && !c.TunnelingEnabled()
```
> We need to detect a direct routing dev (= one which is used to connect K8s Nodes) in order to attach bpf_host when WG is enabled

Don't we need to detect *all* direct routing devices, if only for EKS?

brb replied:
We are interested in attaching `bpf_host` to devices which can be used to communicate among nodes. Detection of such devices (incl. during runtime) should be done by the runtime device detection manager. Does this answer to your concern?

pchaigno replied:
Nop, but I probably didn't ask the right question. I guess my question is: why do we need a special case here in `DirectRoutingDeviceRequired`? I don't see any mention of e.g. the host firewall and yet the host firewall also requires bpf_host to be attached to all native devices.

brb replied:
The auto-device detection is based on two things:

1. IP routes (https://github.com/cilium/cilium/blob/pr/gandro%2Bbrb/wg-host-encryption-v2/pkg/datapath/linux/devices.go#L106).
2. K8s Node IP addr (https://github.com/cilium/cilium/blob/pr/gandro%2Bbrb/wg-host-encryption-v2/pkg/datapath/linux/devices.go#L131).

Usually, the second one is used for communication among nodes. In most cases, the first one should detect the second device, but there are some exclusions (see https://github.com/cilium/cilium/blob/pr/gandro%2Bbrb/wg-host-encryption-v2/pkg/datapath/linux/devices.go#L261). For the encryption we want to be extra defensive, thus we enable the direct routing device detection.

I think we could include the second into the first regardless whether the direct routing device detection is required. I'm going to create an issue for it.

pchaigno replied:
I see. Thanks for the detailed explanation. That indeed sounds like something we'd want to fix longer term.

pchaigno commented on bpf/lib/encap.h:
```diff
@@ -9,6 +9,10 @@
 
+#ifdef ENABLE_WIREGUARD
+#include "lib/wireguard.h"
+#endif
```
Ditto regarding guards.

gandro committed:
docs: Update WireGuard documentation

This adds a new section about node-to-node encryption and removes some
obsolete limitations.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
cmd: Unhide node-encryption flag

It was hidden because it's currently not supported by IPSec. But with
the previous commits, we do now support node-to-node encryption via
WireGuard.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

brb committed:
gh/workflows: Enable node encryption tests

The encryption tests were introduced in
https://github.com/cilium/cilium-cli/pull/1308.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno commented on pkg/datapath/loader/loader.go:
```diff
@@ -236,7 +237,12 @@ func (l *Loader) reloadHostDatapath(ctx context.Context, ep datapath.Endpoint, o
+		if option.Config.AreDevicesRequired() &&
+			// Attaching bpf_host to cilium_wg0 is required for encrypting KPR
+			// traffic. Only ingress prog (aka "from-netdev") is needed to handle
+			// the rev-NAT xlations.
```
Could we also update the comment above the `from-netdev` section in `bpf_host.c`?

pchaigno commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -459,6 +460,17 @@ func finishKubeProxyReplacementInit() error {
+	// it to the stack which drops the packet).
+	if option.Config.EnableNodePort &&
+		option.Config.EnableWireguard && option.Config.EncryptNode {
+		option.Config.AppendDevice(wgTypes.IfaceName)
```
> This commit attaches the bpf_host's "from-netdev" section to the Cilium's WireGuard tunnel netdev ("cilium_wg0").

:scream: Am I the only one worried here?

`from-netdev` implements many features besides reverse NATing. What will happen when host firewall is enabled with Wireguard and policy enforcement starts happening on the ingress (and only ingress) of the wg device?

Loading this here also means we will now perform the local pod delivery from cilium_wg0, no? Doesn't that cause a path asymmetry when endpoint routes are disabled?
1. SYN: lxc -> eth0 -> cilium_wg0 -> eth0
2. SYN+ACK: eth0 -> cilium_wg0 -> lxc

> IMPORTANT: this encrypts KPR traffic only when running in the non-tunneling mode.

Could we avoid loading this here if running in tunneling mode?

brb replied:
Good points! I think we forgot to mention that it's a temporary solution, and we'd like to introduce `from-wireguard` in a follow-up (to keep this PR size sane).

> Loading this here also means we will now perform the local pod delivery from cilium_wg0, no?

I think it should happen only in the case of BPF host routing which is currently mutually exclusive with WG (https://github.com/cilium/cilium/issues/15075).

> Doesn't that cause a path asymmetry when endpoint routes are disabled?

By saying asymmetry, do you mean that the networking stack is bypassed? Or that the eth0 is traversed only once for the reply path?

> Could we avoid loading this here if running in tunneling mode?

Unfortunately, that would break the reply packet path, as replies are coming back through `cilium_wg0`, while requests are forwarded via `cilium_vxlan` (it's a follow up to fix this asymmetry and the encryption issue).

Anyway, created https://github.com/cilium/cilium/issues/23032.

pchaigno replied:
> I think it should happen only in the case of BPF host routing which is currently mutually exclusive with WG (https://github.com/cilium/cilium/issues/15075).

Ah, right. You mean [this check](https://github.com/cilium/cilium/blob/274614fa66b43cf5898fba1164a64b16eafb93b3/bpf/bpf_host.c#L528-L540) (note the comment there is incorrect, we should still be able to perform local delivery when endpoint routes are enabled).

> By saying asymmetry, do you mean that the networking stack is bypassed? Or that the eth0 is traversed only once for the reply path?

I meant the path lxc <> cilium_wg0 was not the same for SYN and SYN+ACK. Not a concern given the above check.

brb replied:
> note the comment there is incorrect

Yikes! I think NO_REDIRECT, BPF host redirect, local redirect etc could be revisited in bpf_host :spaghetti:

pchaigno replied:
Yep. I'd get rid of `NO_REDIRECT` for a start; it's not even used consistently.

brb replied:
I will remove the `NO_REDIRECT`, as I have introduced it :grimacing:

brb replied:
https://github.com/cilium/cilium/pull/23076

brb replied:
For the redirection mess, I think this issue covers it well - https://github.com/cilium/cilium/issues/16166.

pchaigno commented on test/k8s/datapath_configuration.go:
```diff
@@ -564,20 +564,22 @@ var _ = Describe("K8sDatapathConfig", func() {
+				"tunnel":                    "vxlan",
+				"encryption.enabled":        "true",
+				"encryption.type":           "wireguard",
+				"encryption.nodeEncryption": "true",
```
Would there be any benefit to testing node encryption and pod-to-pod encryption separately? Right now if we have a bug in pod-to-pod encryption that is covered by node encryption, we won't notice it. How possible is that?

brb replied:
Good concern. I'm about to drop those tests from the datapath_configuration.go (in a follow-up PR), as they should be covered by the `ci-datapath`.

pchaigno commented on daemon/cmd/daemon_main.go:
```diff
@@ -356,6 +356,9 @@ func initializeFlags() {
 	option.BindEnv(Vp, option.EnableWireguardUserspaceFallback)
 
+	flags.String(option.NodeEncryptionOptOutLabels, defaults.NodeEncryptionOptOutLabels, "Label selector for nodes which will opt-out of node-to-node encryption")
+	option.BindEnv(Vp, option.NodeEncryptionOptOutLabels)
```
*(Commenting here but actually question on the commit description.)*

> This ensures that all Cilium-managed nodes will be able to reach the kube-apiserver running on that node
regardless of encryption status.

Does that mean that some encryption status can cause a persistent loss of connectivity?

pchaigno replied:
The commit and flag descriptions don't explain what opting out of node-to-node encryption means. Does it means excluded_node->node isn't encrypted, or node_excluded->node, or both?

gandro replied:
Both. If the public key of `excluded_node` is stale (and needs to be updated), either direction will be broken, so we opt out from both. The way this is implemented is that:

a) `excluded_node` will set it's own EncryptKey in `CiliumNode` to 0, to ensure `node->excluded_node` is not encrypted by the remote node
b)  `excluded_node` will ignore any EncryptKey of remote `CiliumNode`s, to ensure `excluded_node->node` is not encrypted either.

Note that pod2pod traffic for workloads where the source or target pod are running on `excluded_node` are still encrypted.

We'll revise the docs to make that a bit more clear.

gandro replied:
I missed that, so to give context on why we're doing this:

> Does that mean that some encryption status can cause a persistent loss of connectivity?

If a node looses it's private key (which currently is very possible, e.g. happens if the node reboots, since we don't have a persistent storage for private keys at the moment), it will permanently be locked out of the cluster, until a user manually removes its stale public key from it's CiliumNode CRD. 

The reason this happens is that other nodes (which includes cilium-managed kube-apiservers) will still assume the node's old stale public key. 

The affected node cannot clean up its own CiliumNode CRD, because it cannot establish a connection to the kube-apiserver: The affected node can send packets to the kubeapiserver via plaintext, but the return traffic will still be encrypted with the stale public key, so no TCP session can be established.

pchaigno replied:
Thanks for the great explanations!

pchaigno commented on pkg/defaults/defaults.go:
```diff
@@ -224,6 +224,10 @@ const (
 
+	// NodeEncryptionOptOutLabels contains the label selectors for nodes opting out of
+	// node-to-node encryption
+	NodeEncryptionOptOutLabels = "node-role.kubernetes.io/control-plane"
```
Does this default means that the control plane node is by default excluded from node encryption? Sounds like something we may want to have a big warning for in the docs.

brb replied:
Yep. The docs say:

> By default, Cilium will disable node-to-node encryption on any node
which has the ``node-role.kubernetes.io/control-plane`` label. This is done to
ensure worker nodes are always able to communicate with the kube-apiserver node,
in order to be able to manage their own encryption key. This ensures worker nodes are
never accidentally locked out from the control plane during bootstrap, or
when their private key changes. The label selector matching the control-plane
nodes is configurable via the ``node-encryption-opt-out-labels`` ConfigMap option.

Do you want as to add a note in the beginning of the page?

pchaigno replied:
Oh, I must have missed that yesterday. I would just put it in a Note or "Attention !" box to ensure people are aware that some of their node-to-node traffic is unencrypted by default.

gandro replied:
Good point! We'll revise the docs here to make this a bit more clear (and also document what exactly it means)

gandro replied:
I've rewritten the docs: https://github.com/cilium/cilium/blob/d48de2d736654804e44db3b1d5465ff6b035a307/Documentation/security/network/encryption-wireguard.rst#node-to-node-encryption-beta

Hope this makes it a bit more clear why this is needed (and what the consequences are of disabling it)

pchaigno commented on .github/workflows/conformance-datapath.yaml:
```diff
@@ -266,7 +272,8 @@ jobs:
               --helm-set-string="kubeProxyReplacement=${{ matrix.kpr }}" \
-              --encryption=${{ matrix.encryption }}
+              --encryption=${{ matrix.encryption }} \
+               --helm-set-string="encryption.nodeEncryption=${{ matrix.encryption-node }}"
```
Two commits before, there's this docs change that says we can use a nice `--node-encryption` flag. Is that not possible here because we are using an outdated CLI version?

tklauser replied:
It should be available once we update to cilium-cli v0.12.12: https://github.com/cilium/cilium/pull/23030

brb replied:
I think we should be able to use that by now. Updated.

pchaigno reviewed (changes requested):
Thanks for the great commit history!

Mostly minor comments and nits from me, but I also have a few concerns that may require more changes (in particular about attaching `from-netdev` to the WireGuard device).

chancez commented on pkg/wireguard/agent/agent.go:
```diff
@@ -366,9 +340,21 @@ func (a *Agent) UpdatePeer(nodeName, pubKeyHex string, nodeIPv4, nodeIPv6 net.IP
+			if a.nodeToNodeEncryption {
+				allowedIPs = append(allowedIPs, net.IPNet{
+					IP:   nodeIPv4,
+					Mask: net.CIDRMask(net.IPv4len*8, net.IPv4len*8),
```
Maybe this is only unintuitive to me but I think `32` would be clearer than `net.IPv4len*8`.

gandro replied:
Cilium uses this [pattern elsewhere too](https://github.com/cilium/cilium/blob/48ddc9a494c756a7f6ba03b93bd1715dbd113f99/pkg/ip/ip.go#L18-L19), so I'd say let's keep it for now. We'll eventaully want to switch over to `netip.Addr` anyway, so it will likely be rewritten anyways. But I'd prefer to do that in a follow-up PR, since this one is already large enough.

nebril reviewed (approved):
ci bits lgtm

qmonnet commented on bpf/bpf_lxc.c:
```diff
@@ -539,11 +530,7 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
-		set_encrypt_mark(ctx);
-	else
-#elif !defined(TUNNEL_MODE)
+#if !defined(TUNNEL_MODE)
```
Nit:
```suggestion
#ifndef TUNNEL_MODE
```
+ `/* TUNNEL_MODE */` comment after the `#endif`?

qmonnet commented on bpf/bpf_lxc.c:
```diff
@@ -1115,11 +1084,7 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
-		set_encrypt_mark(ctx);
-	else /* Wireguard and identity mark are mutually exclusive */
-#elif !defined(TUNNEL_MODE)
+#if !defined(TUNNEL_MODE)
```
same

qmonnet commented on bpf/bpf_host.c:
```diff
@@ -1037,6 +1037,7 @@ handle_srv6(struct __ctx_buff *ctx)
  * managed by Cilium (e.g., eth0). This program is only attached when:
  * - the host firewall is enabled, or
  * - BPF NodePort is enabled
+ * - WireGuard's host-to-host encryption is enabled
```
The set of conditions is not super clear when reading just this list: is a `or` missing, or is it `host FW || (NodePort && WG)` or `(host FW || NodePort) && WG)` or...

brb replied:
Good observation, updated (it's OR).

qmonnet commented on pkg/k8s/init.go:
```diff
@@ -181,6 +182,11 @@ func WaitForNodeInformation(ctx context.Context, k8sGetter k8sGetter) error {
 		node.SetLabels(n.Labels)
 
+		if option.Config.NodeEncryptionOptOutLabels.Matches(k8sLabels.Set(n.Labels)) {
+			log.Info("Opting out on node-to-node encryption on this node")
```
I wonder if it's worth adding a note on the origin of that decision? Simply something like:
```suggestion
			log.Info("Opting out on node-to-node encryption on this node as per user configuration")
```

gandro replied:
Good point, I'll add a reference to the `NodeEncryptionOptOutLabels` flag.

qmonnet commented on Documentation/security/network/encryption-wireguard.rst:
```diff
@@ -216,17 +215,73 @@ have WireGuard encryption enabled, i.e. mixed mode is currently not supported.
+  ``node-encryption-opt-out-labels`` ConfigMap option. You may disable this
+  feature by using an empty label selector. Note that disabling this feature
+  requires you to manually update a node's public key in its corresponding
+  ``CiliumNode`` CRD, as the node itself will be unable to do so itself.
```
Reading this paragraph, I found it not entirely clear that `node-encryption-opt-out-labels` defaults to `node-role.kubernetes.io/control-plane`. On my first read, I thought that `node-role.kubernetes.io/control-plane`-labelled nodes could never use node-to-node encryption (OK I missed the “By default” at the beginning, that's my fault), and figured that the `node-encryption-opt-out-labels` was a specifier for _additional_ labels to exclude.

Maybe just me though. Maybe just read this paragraph again, and see if it would make sense to specify that `node-encryption-opt-out-labels` defaults to `node-role.kubernetes.io/control-plane` for the reasons explained above.

gandro replied:
Thanks for the feedback. Given that you are not the first to struggle with the semantics of the flag, it's probably not just you and the text needs some improvements. I'll try to address that.

gandro replied:
I've updated the paragraph slightly to hopefully make it a bit more readable. No fundamental changes, but I hope it still provides a bit more clarity now.

qmonnet replied:
Thank you, it does read better now

qmonnet reviewed (approved):
Looks good to me, with some minor, non-blocking comments. Nice work!

nbusseneau reviewed (approved):
![https://media.giphy.com/media/iP8P6sbQTrmMM/giphy.gif](https://media.giphy.com/media/iP8P6sbQTrmMM/giphy.gif)

qmonnet commented on Documentation/security/network/encryption-wireguard.rst:
```diff
@@ -216,17 +215,81 @@ have WireGuard encryption enabled, i.e. mixed mode is currently not supported.
+  using an empty label selector with that option. Note that doing so is not
+  recommended, as it will require you to always manually update a node's public
+  key in its corresponding ``CiliumNode`` CRD when a worker node's public key
+  changes, as the worker node itself will be unable to do so itself.
```
```suggestion
  changes, given that the worker node will be unable to do so itself.
```

Super-nit: repetitions on `as`, `itself`.

gandro replied:
Thanks, fixed!

tklauser commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -286,6 +286,7 @@ func (n *NodeDiscovery) fillLocalNode() {
 			IP:   node.GetK8sExternalIPv6(),
 		})
 	}
+
```
Nit: stray newline

gandro replied:
Thanks, fixed!

tklauser reviewed (approved):
Thanks for the detailed commit log and the nice graphs in commit 5d234f0f4af9 ("datapath: Change WG integration to support host2host case") :star_struck:
